evolver/hardware/standard/pump.py

```diff
@@ -0,0 +1,112 @@
+        if self.ipp_pumps is True:
+            self.ipp_pumps = list(range(self.slots / 3))
+        elif not self.ipp_pumps:
+            self.ipp_pumps = []
+
+    @property
+    def serial(self):
+        return self.serial_conn or self.evolver.serial
+
+    def set(self, input):
+        self.proposal[input.pump_id] = input
+
+    def commit(self):
+        inputs = copy(self.committed)
+        inputs.update(self.proposal.items())
+        cmd = [b"--"] * self.slots
+        for pump, input in inputs.items():
+            if pump in self.ipp_pumps:
+                # TODO: calibation transform here. The IPP pumps should return array of
+                # hz values per solenoid number - containing 3 of them.
+                hz_a = [input.flow_rate] * 3
+                for solenoid, hz in enumerate(hz_a):
+                    # solenoid is 1-indexed on hardware
+                    cmd[pump * 3 + solenoid] = f"{hz}|{pump}|{solenoid+1}"
+            elif any(pump - i < 3 for i in self.ipp_pumps):
+                raise ValueError(f"pump slot {pump} reserved for IPP pump, cannot address as standard")
+            else:
+                # TODO: calibration transform here. The time_to_pump is the target variable
+                # where pump_interval would presumably be what the pump_interval was set to
+                # during calibration and is fixed post-calibration.
+                time_to_pump, pump_interval = (input.flow_rate, int(input.flow_rate))
+                cmd[pump] = f"{time_to_pump}|{pump_interval}".encode()
+        with self.serial as comm:
+            comm.communicate(SerialData(addr=self.addr, data=cmd))
+        self.committed = inputs
+
+
+class VialIEPump(EffectorDriver):
+    """Vial-based Influx-Efflux Pump driver.
+
+    Each pump in the array is either an influx (to bring liquid to the vial) or
+    efflux (to pull waste out of vial) or spare (unused). Flow rates can be configured
+    separately or as a single rate.
+
+    Configuration parameters ``influx_map`` and ``efflux_map`` can be used to specify
+    explicit mapping between vial and underlying pump_id (of :py:class:`GenericPump`),
+    which by default reserves the first 3rd of slots to influx, and second third to
+    efflux (remaining are spare).
+    """
+
+    class Config(GenericPump.Config):
+        influx_map: dict[int, int] | None = Field(None, description="map of vial to influx pump ID")
+        efflux_map: dict[int, int] | None = Field(None, description="map of vial to efflux pump ID")
+
+    class Input(VialBaseModel):
+        flow_rate_influx: float = Field(description="influx flow rate in ml/s")
+        flow_rate_efflux: float | None = Field(None, description="efflux flow rate in ml/s. Defaults to same as influx")
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        p_slots = self.slots * 3  # influx, efflux, spare (or 3 solenoids for IPP)
+        self._generic_pump = GenericPump(
```

Comment:
In terms of calibration, wondering if this might be better as a separate hardware in evolver and this ones config has a reference to that. Not sure if the calibration would be done per-pump or as a 2-pump unit connected to a vial...